support/cas-server-support-aup-jdbc/src/main/java/org/apereo/cas/aup/JdbcAcceptableUsagePolicyRepository.java

```diff
@@ -23,26 +30,58 @@
                                                final String aupAttributeName,
                                                final DataSource dataSource,
-                                               final String tableName) {
+                                               final AcceptableUsagePolicyProperties properties) {
         super(ticketRegistrySupport, aupAttributeName);
         this.jdbcTemplate = new JdbcTemplate(dataSource);
-        this.tableName = tableName;
+        this.properties = properties;
     }
 
     @Override
     public boolean submit(final RequestContext requestContext, final Credential credential) {
         try {
-            final String sql = String.format("UPDATE %s SET %s=true WHERE username=?", this.tableName, this.aupAttributeName);
-            LOGGER.debug("Executing update query [{}]", sql);
-            return this.jdbcTemplate.update(sql, credential.getId()) > 0;
+            final AcceptableUsagePolicyProperties.Jdbc jdbc = properties.getJdbc();
+            String aupColumnName = properties.getAupAttributeName();
+            if (StringUtils.isNotBlank(jdbc.getAupColumn())) {
+                aupColumnName = jdbc.getAupColumn();
+            }
+            final String sql = String.format(jdbc.getSqlUpdateAUP(), jdbc.getTableName(), aupColumnName, jdbc.getPrincipalIdColumn());
+            final String principalId = determinePrincipalId(requestContext, credential);
+            LOGGER.debug("Executing update query [{}] for principal [{}]", sql, principalId);
+            return this.jdbcTemplate.update(sql, principalId) > 0;
         } catch (final Exception e) {
             LOGGER.error(e.getMessage(), e);
         }
         return false;
     }
+    
+    /**
+     * Extracts principal ID from a principal attribute or the provided credentials.
+     * 
+     * @param requestContext the context
+     * @param credential the credential
+     * @return the principal ID to update the AUP setting in the database for
+     */
+    protected String determinePrincipalId(final RequestContext requestContext, final Credential credential) {
+        if (StringUtils.isBlank(properties.getJdbc().getPrincipalIdAttribute())) {
+            return credential.getId();
+        }
+        @NonNull
+        final Principal principal = WebUtils.getAuthentication(requestContext).getPrincipal();
+        final String pIdAttribName = properties.getJdbc().getPrincipalIdAttribute();
+        if (!principal.getAttributes().containsKey(pIdAttribName)) {
+            throw new IllegalStateException("Principal attribute [" + pIdAttribName + "] cannot be found");
+        }
+        final Object pIdAttributeValue = principal.getAttributes().get(pIdAttribName);
+        final Set<Object> pIdAttributeValues = CollectionUtils.toCollection(pIdAttributeValue);
+        if (pIdAttributeValues.size() != 1) {
+            throw new IllegalStateException("Principal attribute [" + pIdAttribName + "] was found, but its value ["
```

Comment:
Multi-valued attributes should not error out; a warning is likely best while you continue to use the "first" in the list.

Reply:
Okay, but an empty value would still be an error. Please see my changes.